File: packages/coterie-py/src/coterie/adapters/claude_code.py
"""Claude Code CLI adapter."""

import json
from typing import ClassVar

from coterie.adapters.base import AdapterResult, CLIAdapter
from coterie.core.registry import register_adapter


@register_adapter
class ClaudeCodeAdapter(CLIAdapter):
    name: ClassVar[str] = "claude-code"
    # Run on the user's Claude subscription (OAuth session), not a pay-per-token
    # API key — Coterie's coordination LLMs use ANTHROPIC_API_KEY in-process, but
    # the agent CLI should bill against the subscription.
    strip_env: ClassVar[tuple[str, ...]] = ("ANTHROPIC_API_KEY", "ANTHROPIC_AUTH_TOKEN")

    def build_command(self, prompt: str, workdir: str, *, extra: dict) -> list[str]:
        # acceptEdits lets the agent apply file edits without a prompt (the only
        # way it can actually do work headlessly) while still gating riskier
        # tools — safe to run against an isolated workdir. Override via
        # extra["permission_mode"] if a caller wants a different posture.
        permission_mode = extra.get("permission_mode", "acceptEdits")
        cmd = ["claude", "-p", prompt, "--output-format", "json", "--permission-mode", permission_mode]
        if self.model:
            cmd.extend(["--model", self.model])
        return cmd

    def parse_result(self, stdout: str, stderr: str, exit_code: int) -> AdapterResult:
        try:
            payload = json.loads(stdout)
            return AdapterResult(
                stdout=payload.get("result", stdout),
                stderr=stderr,
                exit_code=exit_code,
                cost_estimate_usd=payload.get("total_cost_usd"),
            )
        except json.JSONDecodeError:
            return AdapterResult(stdout=stdout, stderr=stderr, exit_code=exit_code)

File: packages/coterie-js/src/adapters/claudeCode.ts
import { CLIAdapter, type AdapterResult } from "./base.js";
import { parseJsonLoose, parseNdjson } from "../core/json.js";
import { toolHint } from "./stream.js";
import { registerAdapter } from "../core/registry.js";

export class ClaudeCodeAdapter extends CLIAdapter {
  static readonly adapterName = "claude-code";
  // Run on the user's Claude subscription (OAuth session), not a pay-per-token
  // API key — coordination LLMs use ANTHROPIC_API_KEY in-process, but the agent
  // CLI should bill against the subscription.
  static readonly stripEnv = ["ANTHROPIC_API_KEY", "ANTHROPIC_AUTH_TOKEN"];

  buildCommand(prompt: string, _workdir: string, extra: Record<string, unknown> = {}): string[] {
    // Deliberating agents run read-only ('plan' mode: analyse + propose, never
    // edit) — only the finalizer mutates files, with 'acceptEdits' to apply them
    // headlessly while still gating riskier tools. stream-json (which needs
    // --verbose) emits one NDJSON event per step so we can show live progress.
    const permissionMode = extra.readOnly ? "plan" : "acceptEdits";
    const cmd = ["claude", "-p", prompt, "--output-format", "stream-json", "--verbose", "--permission-mode", permissionMode];
    if (this.model) cmd.push("--model", this.model);
    return cmd;
  }

  /** Emit a note for each tool the agent invokes; skip thinking/text/bookkeeping
   *  (the final answer is rendered when the run completes). */
  override streamEvent(line: string): string | null {
    const ev = parseJsonLoose(line);
    if (ev?.type !== "assistant") return null;
    for (const c of ev.message?.content ?? []) {
      if (c?.type === "tool_use") return `→ ${c.name}${toolHint(c.input)}`;
    }
    return null;
  }

  parseResult(stdout: string, stderr: string, exitCode: number): AdapterResult {
    // stdout is an NDJSON stream; the final `result` event carries the answer + cost.
    const events = parseNdjson(stdout);
    const result = events.find((e) => e?.type === "result");
    if (result) {
      const files = events
        .flatMap((e) => (e?.type === "assistant" ? (e.message?.content ?? []) : []))
        .filter((c: any) => c?.type === "tool_use" && /^(Edit|Write|NotebookEdit)$/.test(c.name))
        .map((c: any) => c.input?.file_path)
        .filter(Boolean);
      return {
        stdout: result.result ?? stdout,
        stderr,
        exit_code: exitCode,
        files_changed: [...new Set(files)] as string[],
        duration_s: 0,
        cost_estimate_usd: result.total_cost_usd ?? null,
      };
    }
    // Fallback: not the expected stream (older CLI / plain JSON / prose).
    const loose = parseJsonLoose(stdout);
    return {
      stdout: loose?.result ?? stdout,
      stderr,
      exit_code: exitCode,
      files_changed: [],
      duration_s: 0,
      cost_estimate_usd: loose?.total_cost_usd ?? null,
    };
  }
}

registerAdapter(ClaudeCodeAdapter);
